Remove `Bindable` attribute to reduce binary size impact

vgromfeld opened the pull request:
This PR is related to #3145.
It decreases the toolkit impact on a blank application size by 31Kb.

## PR Type
What kind of change does this PR introduce?
- Bugfix

## What is the current behavior?
Some classes of the toolkit are tagged with the `Bindable` attribute.
This attribute is required for C++ code and is not needed for the C# code which is bindable by default:

From the [MSDN](https://docs.microsoft.com/en-us/uwp/api/Windows.UI.Xaml.Data.BindableAttribute):
> Specifies that a type defined in C++ can be used for binding.
> Apply this attribute to C++-based data classes to enable their use as binding sources. **Common language runtime (CLR) types, including all types defined in C# and Microsoft Visual Basic, are bindable by default**. You can also make a type bindable by implementing ICustomPropertyProvider. For more info, see Data binding in depth.

The `Bindable` attribute is preventing the .Net native compiler to remove the symbol from the generated code.

## What is the new behavior?
I've removed the `Bindable` attribute from the classes that were using it.
This allows me to save **31Kb** on a x86 release build (no directives).

Not a big decrease in size but something we can save without any (big) code change.

## PR Checklist

Please check if your PR fulfills the following requirements:

- [x] Tested code with current [supported SDKs](../readme.md#supported)
- [ ] Pull Request has been submitted to the documentation repository [instructions](..\contributing.md#docs). Link: <!-- docs PR link -->
- [ ] Sample in sample app has been added / updated (for bug fixes / features)
    - [ ] Icon has been created (if new sample) following the [Thumbnail Style Guide and templates](https://github.com/windows-toolkit/WindowsCommunityToolkit-design-assets)
- [ ] Tests for the changes have been added (for bug fixes / features) (if applicable)
- [ ] Header has been added to all new source files (run *build/UpdateHeaders.bat*)
- [x] Contains **NO** breaking changes

Sergio0694 commented:
@vgromfeld Nice find! I've also removed the attribute from the new markup extensions in #3110 😄

michael-hawker commented:
@vgromfeld @Sergio0694 I know it says it's a C++ attribute, and it may be a red herring; but I fell like in the past it's had an effect while generating the XAML Metadata needed to use these elements from the library dynamically in `XamlReader` as per #1961. 

Though maybe it has no effect on the build artifacts, in which case it's safe to remove. I'm not really sure on the best guidance here on how else we're supposed to get the index of available things into the `XamlTypeInfo`. 

This has been a big headache for XAML Studio as I'm constantly having to add manual XAML references in the app itself and entries in the rd.xml file in order to get XamlReader to pick them up at release time. Maybe this is something where it gets easier once .NET 5 hits??? Thoughts?

vgromfeld commented:
From what I'm seeing, the attributes only forces the .Net Native compiler to never remove the symbols from the generated native code (whether the code is used or not). I suspect it automatically adds rules in the `rd.xml` file.

My recommendation would be to be to let the choice be an application choice instead of forcing this for everyone in the toolkit's code. For now, with the limited number of tagged classes, the impact is not important. But if we start tagging everything, we may loose the benefits of the native compilation and generate "bloated" executables.

The `XamlReader` [documentation](https://docs.microsoft.com/en-us/uwp/api/Windows.UI.Xaml.Markup.XamlReader) has the following sentence
> Any custom assemblies referenced in a XAML namespace mapping must already be available to the application.

My initial guess would be that the default `<Assembly Name="*Application*" Dynamic="Required All" />` directive (or something less aggressive like `<Namespace Name="Microsoft.Toolkit" Serialize="Required All" />) should be enough in your case but I may be missing something.

Can you share the errors you were seeing in  Xaml Studio ?

JohnnyWestlake commented:
It definitely used to be required for AttachedProperties to work properly with DependencyProperty bindings in C# UWP XAML, and at one point x:Bind needed them too

michael-hawker commented:
@JohnnyWestlake not sure how relevant that scenario will be for these specific cases; however, I did create a test set of XAML to try for UniformGrid, so I'm just going to build this PR out in Release mode and test that out later today or tomorrow.

michael-hawker commented:
@vgromfeld think @JohnnyWestlake may have had a point. I didn't notice a problem in the Sample app, but our UniformGrid Unit Tests are failing now in the CI results. Can you investigate as they pass on master?

vgromfeld commented:
I've fixed the unit tests. 

They were failing in release because the *uniform grid* tests are using `XamlReader.Load()` to load the control. This is preventing the XAML build task to add the required types to the generated `XamlTypeInfo` file. This link contains interesting background information about this file: [Xaml integration with WinRT and the IXamlMetadataProvider interface](https://jaylee.org/archive/2012/03/08/xaml-integration-with-winrt-and-the-ixamlmetadataprovider-interface.html)

Regular applications using the controls "directly" will not have the problem because the *XAML build task* will properly detect the controls/DP/... and add them to the `XamlTypeInfo`. 

To fix the build pipeline, I've added the dynamically loaded controls in the app resources. Adding them in a unused XAML page will also work. We need this **only** because we are using `XamlReader.Load()`. 

There should be a way to extend the generated `XamlTypeInfo` but this part is not well documented:
- https://cc.bingj.com/cache.aspx?q=IXamlMetadataProvider+custom+implementation&d=4781621615397707&mkt=fr-FR&setlang=fr-FR&w=f4APVHzrdFJ-biX3nKdlLEnrxYrdo_NS
- https://www.tuicool.com/articles/2AvUR3

Adding the `[Bindable]` attribute in the toolkit library code is forcing the *XAML build task* to emit an "assembly XamlTypeInfo" file which is then loaded by the app's `XamlTypeInfo`. See the `OtherProviders` property from `XamlTypeInfo.g.cs`. 
We should not force the generation of this file (and create roots that the .Net Native Compiler cannot discard) since it is probably not relevant for most users who will use the control in a regular way (ie no `XamlReader.Load()`). This is just "bloating" their apps.

michael-hawker commented:
Thanks for the explanation @vgromfeld, that helps clear up a lot of info. I think it's the lack of official documentation that's been the problem in understanding how we resolve #1961. Though does mean we still don't have the best solution for the dynamic access. I was toying with the idea of a T4 template to just dummy use anything we want to access in the Sample App or XAML Studio...

Other than that, I think there's just the build/test issue remaining on the pipeline?

@dotMorten did you ever finish the 2nd part of your `IXamlMetadataProvider` blog?

dotMorten commented:
@michael-hawker There you go making me feel bad. I hardly remember what I was supposed to write in that second blog post :)

dotMorten commented:
> Some classes of the toolkit are tagged with the Bindable attribute.
> This attribute is required for C++ code and is not needed for the C# code which is bindable by default:

That's not entirely correct. If the class you want to bind isn't declared in XAML (directly or indirectly), no metadata will be generated for the type. IMHO removing this would be a breaking change

michael-hawker commented:
> And I’ll leave you hanging here to ponder on that, while I get working on Part 2 of this blogpost where I’ll get back to the custom IXamlMetadataProviders…

We were on the edge of our seats @dotMorten! 😋 

Appreciate you chiming in here and providing some insights. We can move this to the 7.0 branch to be safer in case anyone is relying on that oddity; that release is the primary one themed around optimizations anyway. Thoughts @vgromfeld?

vgromfeld commented:
> We can move this to the 7.0 branch to be safer in case anyone is relying on that oddity; that release is the primary one themed around optimizations anyway. Thoughts @vgromfeld?

Agree to move this to the 7.0 branch. We can then document it properly to explain what should be done when dynamic loading is used.
